src/chapters/unit-testing/unit-testing-in-action.rst

```diff
@@ -0,0 +1,218 @@
+
+Edge Cases
+----------
+
+Recall our definition of **edge case**:
+
+.. pull-quote::
+
+   An edge case is a test case that provides input at the extreme edge of what the unit should be able to handle.
+
+Edge cases can look very different for different units of code. Most of the examples we provided above dealt with numerical edge cases. However, edge cases can also be non-numeric. 
+
+In the case of ``isPalindrome``, the most obvious edge case would be that of the empty string, ``""``. This is the smallest possible string that we can use when calling ``isPalindrome``. Not only is it the smallest, but it is essentially *different* from the next longest string, ``"a"``---one has characters and one doesn't. 
+
+Should the empty string be considered a palindrome? That decision is up to us, the programmer, and there is not right or wrong answer. In our case, we have decided to take a very literal definition of the term "palindrome" by considering case and whitespace. In other words, our definition says that a string is a palindrome exactly when it equals its reverse. Since the reverse of ``""`` is also ``""``, it makes sense to consider the emtpy string a palindrome.
+
+Let's add this test case to our spec:
+
+.. sourcecode:: js
+
+   it("should consider the empty string a palindrome", function(){
+      assert.strictEqual(isPalindrome(""), true);
+   });
+
+Now run the tests, which should all pass.
+
+You might think that another edge case is that of the longest possible palindrome. Such a palindrome would be as long as the longest possible string in JavaScript. This case is not worth considering for a couple of reasons:
+
+- The length of the longest string `can vary across different JavaScript implementations <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#Description>`_.
+- The most recent JavaScript specification, ES2016, states that the maximum allowed length of a string should be 2 :sup:`53` - 1 characters. This is a LOT of characters, and it is unrealistic to expect that our function will ever be given such a string.
+
+Toward a Better Testing Workflow
+--------------------------------
+
+In this case, we had a well-written function to write tests for, so it was straightforward to create tests that pass. Most situations will not be this simple. You will often find yourself writing tests for code only to discover a bug, forcing you to go back and update your code. That's okay! This is precisely what tests are for.
+
+The workflow for this situation is:
+
+#. Write code
+#. Write tests
+#. Fix any bugs found while testing
+
+The rest of the chapter focuses on a programming technique that allows you to completly *eliminate* the third step, by reversing the order of the first two:
+
+#. Write tests
+#. Write code
+
+As you will soon learn, writing your tests *before* writing your code is a great way to enhance your programming efficiency and quality.
```

Comment:
Word removal: "As you will soon learn, writing your tests *before* the code is a great way to enhance your programming efficiency and quality."